### PR TITLE
Implement FRI Prover

### PR DESCRIPTION
**Pull request by varunthakore**
Implements FRI prover and correctness tests corresponding to the following Stone library:

https://github.com/starkware-libs/stone-prover/tree/main/src/starkware/fri

This PR has dependence on #18 and #19.

### CODE REVIEW

File: fri/src/prover.rs [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+            committed_layers,
+            mont_r,
+        }
+    }
+
+    fn create_next_fri_layer(
+        channel: &mut FSProverChannel<F, P, W>,
+        layer: Arc<dyn FriLayer<F, Radix2EvaluationDomain<F>>>,
+        fri_step: usize,
+        _i: usize,
```
Comment (rot256) [inline fri/src/prover.rs:56]: Remove the `_i` argument?

File: fri/src/prover.rs [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+
+pub fn to_mont_repr<F: PrimeField>(segment: &[Vec<F>], mont_r: F) -> Vec<Vec<F>> {
+    let mut result: Vec<Vec<F>> = Vec::with_capacity(segment.len());
+
+    for row in segment.iter() {
+        let mut new_row = Vec::with_capacity(row.len());
+        for element in row.iter() {
+            new_row.push(*element * mont_r);
+        }
+        result.push(new_row);
```
Comment (rot256) [inline fri/src/prover.rs:201]:
```rust
result.push(row.iter().cloned().map(|e| e * mont_r).collect())
```

**Review by rot256 — APPROVED** [review]
Not a lot of comments from me. Looks good!

**Comment by varunthakore** [comment]
Closing this PR since #24 is merged.

**Pull request closed without merging**